### ISSUE TITLE
How does LSTM  iterate through a sequence using `states` when `stateful==False`?

### ISSUE THREAD

**Issue by dragonhyq** [fchollet/keras#3445]
Hi! I am very new in Theano and Keras, apology for my ignorance. 

When `stateful==False`, the state is `[None, None]`. but the  [step](https://github.com/fchollet/keras/blob/master/keras/layers/recurrent.py#L788-L789) function uses the state, ie `h_tm1` and `c_tm1`. 

Keras' LSTM certainly works when `stateful=False`, using the [code](https://gist.github.com/philipperemy/7104d9072d7f9fde55abe1c76f8a33ed) from [Philippe Remy's blog](http://philipperemy.github.io/keras-stateful-lstm/) with settings `USE_SEQUENCES = False` and `USE_STATELESS_MODEL = True`: the model quickly achieves 1.00 accuracy. That means the state cannot be `[None, None]`, otherwise the model wouldn't be able to learn. Would it? 

I do not understand. Could someone explain this? 

By the way I am trying to implement the encoder-decoder structure by Cho and Sutskever, and I am at the phase of understanding "keras/layers/recurrent.py".

Many thanks in advance!

**Comment by mbollmann** [comment]
`stateful=True` means that states are preserved **across batches**, while `stateful=False` resets the states for each batch and each sequence. It does *not* mean the model doesn't use states.

**Comment by dragonhyq** [comment]
Thanks! I see. So far my understanding it that the within-batch states are used in `scan()`, and only `initial_states` is needed for `scan()` to run.

I think the way suggested #2995 would work, but I think I need to know how the entire pipeline of keras work (starting with reading keras.models.Sequential) so that I can subclass LSTM/GRU correctly. Any better suggestion is most certainly appreciated!
